source/multigrid/mg_tools.cc

```diff
@@ -1201,31 +1201,99 @@ namespace MGTools
     std::vector<std::set<types::global_dof_index> > &boundary_indices,
     const ComponentMask &component_mask)
   {
-    // if for whatever reason we were
-    // passed an empty map, return
-    // immediately
-    if (function_map.size() == 0)
-      return;
+    Assert (boundary_indices.size() == dof.get_triangulation().n_global_levels(),
+            ExcDimensionMismatch (boundary_indices.size(),
+                                  dof.get_triangulation().n_global_levels()));
 
-    const unsigned int n_levels = dof.get_triangulation().n_global_levels();
+    std::set<types::boundary_id> boundary_indicators;
+    typename std::map<types::boundary_id, const Function<dim>* >::const_iterator it
+      = function_map.begin();
+    for (; it!=function_map.end(); ++it)
+      boundary_indicators.insert(it->first);
 
-    (void)n_levels;
+    std::vector<IndexSet>  boundary_indexset;
+    make_boundary_list(dof, boundary_indicators, boundary_indexset, component_mask);
+    for (unsigned int i=0; i<dof.get_triangulation().n_global_levels(); ++i)
+      {
+        IndexSet::ElementIterator it_element = boundary_indexset[i].begin();
+        for (; it_element != boundary_indexset[i].end(); ++it_element)
+          boundary_indices[i].insert(*it_element);
+      }
+  }
 
 
-    const unsigned int n_components = DoFTools::n_components(dof);
-    const bool          fe_is_system = (n_components != 1);
+  template <int dim, int spacedim>
+  void
+  make_boundary_list(const DoFHandler<dim,spacedim> &dof,
+                     const typename FunctionMap<dim>::type &function_map,
+                     std::vector<IndexSet> &boundary_indices,
+                     const ComponentMask &component_mask)
+  {
+    Assert (boundary_indices.size() == dof.get_triangulation().n_global_levels(),
+            ExcDimensionMismatch (boundary_indices.size(),
+                                  dof.get_triangulation().n_global_levels()));
+
+    std::set<types::boundary_id> boundary_indicators;
+    typename std::map<types::boundary_id, const Function<dim>* >::const_iterator it = function_map.begin();
+    for (; it!=function_map.end(); ++it)
+      boundary_indicators.insert(it->first);
+
+    make_boundary_list (dof, boundary_indicators, boundary_indices, component_mask);
+  }
+
+
+  template <>
+  void
+  make_boundary_list(
+    const DoFHandler<1,1> &,
+    const std::set<types::boundary_id> &,
+    std::vector<IndexSet> &,
+    const ComponentMask &)
+  {
+    Assert(false, ExcNotImplemented());
+  }
```

Comment:
I see. I added the function in #2853 just now. If you want, wait till that is merged and rebase -- you may then be able to remove the specializations. Or we can take care of that as a follow-up.